src/plugins/flow.js

```diff
@@ -21,6 +21,46 @@ pp.flowParseTypeInitialiser = function (tok, allowLeadingPipeOrAnd) {
   return type;
 };
 
+pp.flowParsePredicate = function() {
+  let node = this.startNode();
+  let moduloLoc = this.state.startLoc;
+  let moduloPos = this.state.start;
+  this.expect(tt.modulo);
+  let checksLoc = this.state.startLoc;
+  this.expectContextual("checks");
+  // Force '%' and 'checks' to be adjacent
+  if (moduloLoc.line !== checksLoc.line || moduloLoc.column !== checksLoc.column - 1) {
+    this.unexpected(moduloPos);
+  }
+  if (this.match(tt.parenL)) {
+    this.next();
+    node.expression = this.parseExpression();
+    this.expect(tt.parenR);
+    return this.finishNode(node, "DeclaredPredicate");
+  } else {
+    return this.finishNode(node, "InferredPredicate");
+  }
+};
+
+pp.flowParseTypeAndPredicateInitialiser = function () {
+  let oldInType = this.state.inType;
+  this.state.inType = true;
```

Comment:
It looks like `this.state.inType` needs to be `true` for `this.expect(tt.colon)` to succeed (I tried to move it further down, but other tests failed). So, `oldInType` needs to be recorded in the beginning in order to be restored before `this.flowParsePredicate()` is called (note that `flowParsePredicate` possibly parses expressions).

Bottom line is I wasn't able to perform the optimization you suggested.